{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "dee1f814-9b74-4acc-9f44-a58dbe6f963e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dash\n",
    "from dash import dcc, html, Input, Output, State\n",
    "import dash.dash_table\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from plotly.colors import sample_colorscale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "53138872-dd14-49fc-ba38-c06c54f06af0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the merged data\n",
    "df = pd.read_csv(\"../results/dash_expression_data.csv\", na_values=[\"NA\", \"NaN\", \"\"])\n",
    "\n",
    "# Clean for safety\n",
    "df['gene'] = df['gene'].astype(str)\n",
    "df['condition'] = df['condition'].astype(str)\n",
    "# Convert to numeric, ensuring float64 dtype\n",
    "df[\"log2FoldChange\"] = pd.to_numeric(df[\"log2FoldChange\"], errors=\"coerce\", downcast=None)\n",
    "df['FDR'] = pd.to_numeric(df['FDR'], errors='coerce', downcast=None)\n",
    "#df[\"FDR\"] = pd.to_numeric(df[\"FDR\"], errors=\"coerce\")\n",
    "\n",
    "# Set up gene list\n",
    "genes = sorted(df['gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9b398856-fc32-4f48-8b83-acf14b9edc36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   gene sample  expression     ID condition          Run  \\\n",
      "0                PRDM16   TLE1    3.321570   TLE1       TLE  SRR10446605   \n",
      "1                PRDM16   TLE2    3.558254   TLE2       TLE  SRR10446608   \n",
      "2                PRDM16   TLE3    3.117902   TLE3       TLE  SRR10446611   \n",
      "3                PRDM16  Ctrl1    3.471156  Ctrl1        TL  SRR10446596   \n",
      "4                PRDM16  Ctrl2    3.739109  Ctrl2        TL  SRR10446599   \n",
      "...                 ...    ...         ...    ...       ...          ...   \n",
      "159337  ENSG00000300815   TLE2    3.434051   TLE2       TLE  SRR10446608   \n",
      "159338  ENSG00000300815   TLE3    2.563037   TLE3       TLE  SRR10446611   \n",
      "159339  ENSG00000300815  Ctrl1    3.650406  Ctrl1        TL  SRR10446596   \n",
      "159340  ENSG00000300815  Ctrl2    2.987791  Ctrl2        TL  SRR10446599   \n",
      "159341  ENSG00000300815  Ctrl3    2.973240  Ctrl3        TL  SRR10446602   \n",
      "\n",
      "        sizeFactor  log2FoldChange  FDR  \n",
      "0         0.681726             NaN  NaN  \n",
      "1         1.043081             NaN  NaN  \n",
      "2         0.983671             NaN  NaN  \n",
      "3         0.732454             NaN  NaN  \n",
      "4         1.236764             NaN  NaN  \n",
      "...            ...             ...  ...  \n",
      "159337    1.043081             NaN  NaN  \n",
      "159338    0.983671             NaN  NaN  \n",
      "159339    0.732454             NaN  NaN  \n",
      "159340    1.236764             NaN  NaN  \n",
      "159341    1.698848             NaN  NaN  \n",
      "\n",
      "[159342 rows x 9 columns]\n"
     ]
    }
   ],
   "source": [
    "print(df)\n",
    "#df['FDR'].min(skipna=True)\n",
    "#df['FDR'].max(skipna=True)\n",
    "#print(df.columns.tolist())\n",
    "#print(df[\"FDR\"].head(10))\n",
    "#print(df[\"FDR\"].dtype)\n",
    "#range(df['FDR'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "a285bd6a-d7a1-4f6a-9585-331951f1005d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8050/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x1c4df9ad630>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Dash app setup\n",
    "app = dash.Dash(__name__)\n",
    "\n",
    "default_gene = df.loc[df['log2FoldChange'].abs().idxmax(), 'gene']\n",
    "\n",
    "#Viridis color palette\n",
    "viridis_colors = sample_colorscale(\"Viridis\", [0.75, 0.25])\n",
    "\n",
    "app.layout = html.Div([\n",
    "  html.H2(\"Gene Expression Dashboard\", style={\"textAlign\": \"center\"}),\n",
    "  dcc.Checklist(\n",
    "    id=\"volcano-toggle\",\n",
    "    options=[{\"label\": \" Show volcano plot\", \"value\": \"show\"}],\n",
    "    value=[],\n",
    "    style={\"margin\": \"10px 0\"}\n",
    "),\n",
    "  html.Div(id=\"volcano-container\"),\n",
    "\n",
    "  dcc.Dropdown(\n",
    "    id=\"gene-dropdown\",\n",
    "    options=[{\"label\": g, \"value\": g} for g in sorted(df[\"gene\"].dropna().unique())],\n",
    "    # Set a specific gene of interest as the default gene\n",
    "    #value=\"LGR6\",\n",
    "    value=default_gene,\n",
    "    style={\"width\": \"200px\"}\n",
    "  ),\n",
    "  dcc.Graph(id=\"expression-plot\",\n",
    "           style={\"width\": \"100%\",\n",
    "                  \"margin\": \"0 auto\"}\n",
    "           ),\n",
    "  dcc.Checklist(\n",
    "    id='sig-only-toggle',\n",
    "    options=[{'label': ' Show only significant genes (FDR < 0.05)', 'value': 'sig'}],\n",
    "    value=[],  # empty = show all by default\n",
    "    style={\"margin\": \"10px\"}\n",
    "),\n",
    "  html.Button(\"Download graph data (CSV)\", id=\"download-btn\", n_clicks=0, style={\"margin\": \"20px\"}),\n",
    "  dcc.Download(id=\"download-data\"),\n",
    "  html.Div(id=\"de-info-text\", style={\"margin\": \"10px\", \"textAlign\": \"center\", \"fontSize\": \"14px\"}),\n",
    "  dash.dash_table.DataTable(\n",
    "    id='expression-table',\n",
    "    style_table={'overflowX': 'auto', 'margin': '20px auto', 'width': '70%'},\n",
    "    style_cell={'textAlign': 'left',\n",
    "                'padding': '5px',\n",
    "                'fontFamily': 'sans-serif',\n",
    "                'fontSize': '14px'\n",
    "               },\n",
    "    style_data_conditional=[\n",
    "        {\n",
    "            'if': {'column_id': 'expression'},\n",
    "            'textAlign': 'right',\n",
    "            'fontFamily': 'monospace'\n",
    "        }\n",
    "    ],\n",
    "    style_header={'fontWeight': 'bold'},\n",
    "    page_size=10\n",
    "  )\n",
    "])\n",
    "\n",
    "@app.callback(\n",
    "    Output(\"volcano-container\", \"children\"),\n",
    "    Input(\"volcano-toggle\", \"value\")\n",
    ")\n",
    "def show_volcano(toggle_vals):\n",
    "    if \"show\" not in toggle_vals:\n",
    "        return None\n",
    "\n",
    "    volcano_df = df.copy()\n",
    "    volcano_df[\"-log10FDR\"] = -np.log10(volcano_df[\"FDR\"])\n",
    "    volcano_df[\"Significant\"] = volcano_df[\"FDR\"] < 0.05\n",
    "\n",
    "    fig = px.scatter(\n",
    "        volcano_df,\n",
    "        x=\"log2FoldChange\",\n",
    "        y=\"-log10FDR\",\n",
    "        color=\"Significant\",\n",
    "        hover_name=\"gene\",\n",
    "        custom_data=[\"gene\"],  # needed for click-to-select\n",
    "        color_discrete_map={True: \"crimson\", False: \"gray\"}\n",
    "    )\n",
    "\n",
    "    fig.update_traces(marker=dict(size=6, opacity=0.7))\n",
    "    fig.update_layout(\n",
    "        title=\"Volcano Plot\",\n",
    "        height=450,\n",
    "        margin=dict(t=50, l=40, r=20, b=40),\n",
    "        xaxis_title=\"log₂ Fold Change\",\n",
    "        yaxis_title=\"-log₁₀(FDR)\",\n",
    "        plot_bgcolor=\"white\"\n",
    "    )\n",
    "    fig.update_xaxes(linecolor=\"black\", zeroline=True, zerolinewidth=1)\n",
    "    fig.update_yaxes(linecolor=\"black\")\n",
    "\n",
    "    return dcc.Graph(id=\"volcano-plot\", figure=fig)\n",
    "                                                                    \n",
    "@app.callback(\n",
    "  Output(\"gene-dropdown\", \"options\"),\n",
    "  Output(\"gene-dropdown\", \"value\"),\n",
    "  Input(\"sig-only-toggle\", \"value\")\n",
    ")\n",
    "\n",
    "def update_gene_list(toggle_values):\n",
    "    if \"sig\" in toggle_values:\n",
    "        filtered = df[df[\"FDR\"] < 0.05]\n",
    "    else:\n",
    "        filtered = df\n",
    "\n",
    "    # Remove NAs and duplicates\n",
    "    gene_list = sorted(filtered[\"gene\"].dropna().unique())\n",
    "\n",
    "    # Default gene: max abs log2FC\n",
    "    if len(gene_list) == 0:\n",
    "        return [], None\n",
    "    default_gene = (\n",
    "        filtered.loc[filtered[\"log2FoldChange\"].abs().idxmax(), \"gene\"]\n",
    "        if \"log2FoldChange\" in filtered.columns else gene_list[0]\n",
    "    )\n",
    "\n",
    "    options = [{\"label\": g, \"value\": g} for g in gene_list]\n",
    "    \n",
    "    return options, default_gene\n",
    "\n",
    "\n",
    "@app.callback(\n",
    "  Output(\"expression-plot\", \"figure\"),\n",
    "  Input(\"gene-dropdown\", \"value\")\n",
    ")\n",
    "\n",
    "def update_plot(selected_gene):\n",
    "    #Define dataset for the plot\n",
    "    data = df[df[\"gene\"] == selected_gene].copy()\n",
    "\n",
    "    #Define FDR and LFC text\n",
    "    padj = data[\"FDR\"].iloc[0]\n",
    "    lfc = data[\"log2FoldChange\"].iloc[0]\n",
    "    padj_text = f\"FDR = {padj:.2e}\" if pd.notnull(padj) else \"No DE result\"\n",
    "\n",
    "    fig = go.Figure()\n",
    "\n",
    "    #Convert conditions from catagorical to numeric to plot boxplots and jittered datapoints on the same field\n",
    "    conditions = [\"TL\", \"TLE\"]  # TL = Control, TLE = Epilepsy\n",
    "    cond_to_x = {cond: i for i, cond in enumerate(conditions)}\n",
    "    #Add condition labels.\n",
    "    cond_labels = {\n",
    "    \"TL\": \"Control\",\n",
    "    \"TLE\": \"Epilepsy\"\n",
    "    }\n",
    "    \n",
    "    for i, cond in enumerate(conditions):\n",
    "        subset = data[data[\"condition\"] == cond]\n",
    "        x_val = cond_to_x[cond]\n",
    "        \n",
    "        #Boxplot\n",
    "        fig.add_trace(go.Box(\n",
    "            y=subset[\"expression\"],\n",
    "            x=[x_val] * len(subset),\n",
    "            name=cond,\n",
    "            marker_color=\"white\", #fill color\n",
    "            fillcolor=\"white\", # explicitly set\n",
    "            line=dict(color=\"black\"), #line color\n",
    "            width=0.25,\n",
    "            boxpoints=\"outliers\",\n",
    "            showlegend=False\n",
    "        ))\n",
    "\n",
    "        # Dot plot (viridis-colored with jitter)\n",
    "        x_jittered = [x_val + np.random.uniform(-0.15, 0.15) for _ in range(len(subset))]\n",
    "\n",
    "        #Scatter points\n",
    "        fig.add_trace(go.Scatter(\n",
    "            y=subset[\"expression\"],\n",
    "            x=x_jittered,\n",
    "            mode=\"markers\",\n",
    "            marker=dict(\n",
    "                size=12, \n",
    "                color=viridis_colors[i % len(viridis_colors)]\n",
    "            ),\n",
    "            hovertext=subset[\"sample\"],\n",
    "            showlegend=False\n",
    "        ))\n",
    "\n",
    "        #Add significance bar for DE genes\n",
    "        # Check if gene is significant\n",
    "        fdr = data[\"FDR\"].iloc[0]\n",
    "        lfc = data[\"log2FoldChange\"].iloc[0]\n",
    "        is_significant = pd.notnull(fdr) and fdr < 0.05\n",
    "        #y_max = data[\"expression\"].max()\n",
    "        #bar_y = y_max * 1.2  # pad above the data\n",
    "\n",
    "        if is_significant:\n",
    "            y_max = data[\"expression\"].max()\n",
    "            bar_y = y_max * 1.05\n",
    "        \n",
    "            # Horizontal line\n",
    "            fig.add_trace(go.Scatter(\n",
    "                x=[0, 1],\n",
    "                y=[bar_y, bar_y],\n",
    "                mode=\"lines\",\n",
    "                line=dict(color=\"black\", width=1),\n",
    "                showlegend=False,\n",
    "                hoverinfo=\"skip\"\n",
    "            ))\n",
    "        \n",
    "            # Vertical line 0\n",
    "            fig.add_trace(go.Scatter(\n",
    "                x=[0, 0],\n",
    "                y=[(bar_y*0.98), bar_y],\n",
    "                mode=\"lines\",\n",
    "                line=dict(color=\"black\", width=1),\n",
    "                showlegend=False,\n",
    "                hoverinfo=\"skip\"\n",
    "            ))\n",
    "\n",
    "            # Vertical line 1\n",
    "            fig.add_trace(go.Scatter(\n",
    "                x=[1, 1],\n",
    "                y=[(bar_y*0.98), bar_y],\n",
    "                mode=\"lines\",\n",
    "                line=dict(color=\"black\", width=1),\n",
    "                showlegend=False,\n",
    "                hoverinfo=\"skip\"\n",
    "            ))\n",
    "\n",
    "    fig.update_layout(\n",
    "        title=f\"<i>{selected_gene}</i>\",\n",
    "        yaxis_title=\"Expression (log scale)\",\n",
    "        title_x=0.5,\n",
    "        font=dict(size=12),\n",
    "        plot_bgcolor=\"white\",\n",
    "        xaxis=dict(\n",
    "        range=[-0.5, 1.5],  # tighter than default for 2 categories\n",
    "        ),\n",
    "        annotations=[\n",
    "            dict(\n",
    "                text=padj_text,\n",
    "                x=0.5,\n",
    "                y=1.15,\n",
    "                xref=\"paper\",\n",
    "                yref=\"paper\",\n",
    "                showarrow=False,\n",
    "                font=dict(size=12),\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    fig.update_yaxes(type=\"log\", linecolor=\"black\")\n",
    "    fig.update_xaxes(\n",
    "        linecolor=\"black\",\n",
    "        tickangle=-45,\n",
    "        tickmode=\"array\",\n",
    "        tickvals=list(cond_to_x.values()),\n",
    "        ticktext=[cond_labels[cond] for cond in conditions],\n",
    "        type=\"linear\"\n",
    "    )\n",
    "\n",
    "    return fig\n",
    "\n",
    "@app.callback(\n",
    "    Output(\"download-data\", \"data\"),\n",
    "    Input(\"download-btn\", \"n_clicks\"),\n",
    "    State(\"gene-dropdown\", \"value\"),\n",
    "    prevent_initial_call=True\n",
    ")\n",
    "def download_table(n_clicks, selected_gene):\n",
    "    if selected_gene is None:\n",
    "        return dash.no_update\n",
    "\n",
    "    data_subset = df[df[\"gene\"] == selected_gene].copy()\n",
    "\n",
    "    columns_to_export = [\"sample\", \"condition\", \"expression\"]\n",
    "    if \"FDR\" in data_subset.columns and \"log2FoldChange\" in data_subset.columns:\n",
    "        columns_to_export += [\"log2FoldChange\", \"FDR\"]\n",
    "\n",
    "    return dcc.send_data_frame(\n",
    "        data_subset[columns_to_export].to_csv,\n",
    "        filename=f\"{selected_gene}_expression_table.csv\",\n",
    "        index=False\n",
    "    )\n",
    "\n",
    "@app.callback(\n",
    "    Output(\"de-info-text\", \"children\"),\n",
    "    Input(\"gene-dropdown\", \"value\")\n",
    ")\n",
    "def update_de_info(selected_gene):\n",
    "    if selected_gene is None:\n",
    "        return \"\"\n",
    "\n",
    "    gene_data = df[df[\"gene\"] == selected_gene]\n",
    "\n",
    "    if \"FDR\" in gene_data.columns and \"log2FoldChange\" in gene_data.columns:\n",
    "        fdr_val = gene_data[\"FDR\"].iloc[0]\n",
    "        lfc_val = gene_data[\"log2FoldChange\"].iloc[0]\n",
    "\n",
    "        if pd.notnull(fdr_val) and pd.notnull(lfc_val):\n",
    "            return f\"log₂FC: {lfc_val:.3f} | FDR: {fdr_val:.2e}\"\n",
    "        else:\n",
    "            return \"No differential expression data available for this gene.\"\n",
    "    else:\n",
    "        return \"DE information not available.\"\n",
    "\n",
    "@app.callback(\n",
    "    Output(\"expression-table\", \"data\"),\n",
    "    Output(\"expression-table\", \"columns\"),\n",
    "    Input(\"gene-dropdown\", \"value\")\n",
    ")\n",
    "def update_table(selected_gene):\n",
    "    if selected_gene is None:\n",
    "        return [], []\n",
    "\n",
    "    data_subset = df[df[\"gene\"] == selected_gene].copy()\n",
    "    #Round expression data to 3 significant digits\n",
    "    data_subset[\"expression\"] = data_subset[\"expression\"].apply(lambda x: f\"{x:.4g}\")\n",
    "\n",
    "    #Choose what to show in the table\n",
    "    columns_to_show = [\"sample\", \"condition\", \"expression\"]\n",
    "    #Optionally include FDR, log2FC, or any metadata\n",
    "    #if \"FDR\" in data_subset.columns and \"log2FoldChange\" in data_subset.columns:\n",
    "    #    data_subset[\"FDR\"] = data_subset[\"FDR\"].round(4)\n",
    "    #    data_subset[\"log2FoldChange\"] = data_subset[\"log2FoldChange\"].round(3)\n",
    "    #    columns_to_show += [\"log2FoldChange\", \"FDR\"]\n",
    "    #Replace columns for readability\n",
    "    column_headers = {\n",
    "    \"sample\": \"Sample\",\n",
    "    \"condition\": \"Condition\",\n",
    "    \"expression\": \"Expression (log scale)\"\n",
    "    #\"log2FoldChange\": \"log₂ Fold Change\",\n",
    "    #\"FDR\": \"FDR (q-value)\"\n",
    "    }\n",
    "    \n",
    "    table_data = data_subset[columns_to_show].to_dict(\"records\")\n",
    "    table_columns = [{\"name\": column_headers[col], \"id\": col} for col in columns_to_show]\n",
    "\n",
    "    return table_data, table_columns\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "  app.run(debug=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "f51ecbc0-ca3c-4c88-b8a6-f3ce5661ae76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['rgb(96, 200, 96)', 'rgb(59, 81, 138)']\n"
     ]
    }
   ],
   "source": [
    "viridis_colors = sample_colorscale(\"Viridis\", [0.75, 0.25])\n",
    "print(viridis_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "043bbb1b-7bce-4552-b827-4ea726a75060",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(59, 82, 139)\n"
     ]
    }
   ],
   "source": [
    "def hex_to_rgb(hex_color):\n",
    "    hex_color = hex_color.lstrip('#')\n",
    "    return tuple(int(hex_color[i:i + 2], 16) for i in (0, 2, 4))\n",
    "\n",
    "# Example usage:\n",
    "hex_color = \"#3B528B\"\n",
    "rgb_color = hex_to_rgb(hex_color)\n",
    "print(rgb_color)  # Output: (255, 0, 0)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
